internal/cephfs/controllerserver.go

```diff
@@ -36,25 +41,93 @@ type ControllerServer struct {
 	// A map storing all volumes with ongoing operations so that additional operations
 	// for that same volume (as defined by VolumeID/volume name) return an Aborted error
 	VolumeLocks *util.VolumeLocks
+
+	// A map storing all snapshots with ongoing operations so that additional operations
+	// for that same snapshot (as defined by SnapshotID/snapshot name) return an Aborted error
+	SnapshotLocks *util.VolumeLocks
+
+	// A map storing all volumes/snapshots with ongoing operations.
+	OperationLocks *util.OperationLock
 }
 
 // createBackingVolume creates the backing subvolume and on any error cleans up any created entities.
-func (cs *ControllerServer) createBackingVolume(ctx context.Context, volOptions *volumeOptions, vID *volumeIdentifier, secret map[string]string) error {
-	cr, err := util.NewAdminCredentials(secret)
-	if err != nil {
-		return status.Error(codes.InvalidArgument, err.Error())
+func (cs *ControllerServer) createBackingVolume(
+	ctx context.Context,
+	volOptions,
+	parentVolOpt *volumeOptions,
+
+	vID,
+	pvID *volumeIdentifier,
+	sID *snapshotIdentifier,
+	cr *util.Credentials) error {
+	var err error
+	if sID != nil {
+		if err = cs.OperationLocks.GetRestoreLock(sID.SnapshotID); err != nil {
+			klog.Error(util.Log(ctx, err.Error()))
+			return status.Error(codes.Aborted, err.Error())
+		}
+		defer cs.OperationLocks.ReleaseRestoreLock(sID.SnapshotID)
+
+		err = createCloneFromSnapshot(ctx, parentVolOpt, volOptions, vID, sID, cr)
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
```

Comment:
return `err` as it is. as we need to compare cloneinprogress

Reply:
yep. done.